lib/ansible/plugins/action/gather_facts.py

```
@@ -0,0 +1,86 @@
+
+
+class ActionModule(ActionBase):
+
+    def _get_module_args(self, fact_module, task_vars):
+        mod_args = task_vars.get('ansible_facts_modules', {}).get(fact_module, {})
+        if isinstance(mod_args, MutableMapping):
+            mod_args.update(self._task.args.copy())
+        else:
+            mod_args = self._task.args.copy()
+
+        if fact_module != 'setup':
+            mod_args.pop('gather_subset', None)
+
+        return mod_args
+
+    def run(self, tmp=None, task_vars=None):
+        ''' handler for package operations '''
+
+        self._supports_check_mode = True
+
+        result = super(ActionModule, self).run(tmp, task_vars)
+        result['ansible_facts'] = {}
+
+        jobs = {}
+        modules = C.config.get_config_value('FACTS_MODULES', variables=task_vars)
+        if len(modules) > 1:
+            for fact_module in modules:
+
+                mod_args = self._get_module_args(fact_module, task_vars)
+                self._display.vvvv("Running %s" % fact_module)
+                jobs[fact_module] = (self._execute_module(module_name=fact_module, module_args=mod_args, task_vars=task_vars, wrap_async=True))
+
+            failed = {}
+            skipped = {}
+            while jobs:
+                for module in jobs:
+                    poll_args = {'jid': jobs[module]['ansible_job_id'], '_async_dir': os.path.dirname(jobs[module]['results_file'])}
+                    res = self._execute_module(module_name='async_status', module_args=poll_args, task_vars=task_vars, wrap_async=False)
+                    if 'finished' in res:
+                        if res.get('failed', False):
+                            failed[module] = res.get('msg')
+                        elif res.get('skipped', False):
+                            skipped[module] = res.get('msg')
+                        else:
+                            result = combine_vars(result, {'ansible_facts': res.get('ansible_facts', {})})
```

Are we OK silently masking collisions between facts modules (last writer wins)? I think I'm OK with that (again, pretty minor corner case), but maybe a warning would be better (then again, somebody's going to want merge_hash to work at some point, and ick).

---

it works the same as currently when you execute them yourself, last one updating wins

---

The one difference being that exec order is not guaranteed since the source driver is a dict. Even if it were an OrderedDict and we tried to preserve the key order (good luck), we're updating the results when the tasks finish, not in the order they were run in. 

I don't think this is a showstopper, but might be worth a note in the docs or something that ordering isn't guaranteed, and that any colliding keys are overwritten in an undefined order.

---

agreed, on note, though i don't really expect multiple modules to return the same keys with diff values.